File: apps/mobile/lib/domain/models/course_package_manifest.dart
// Course Package Manifest Model — VSP Mobile App
//
// The manifest contract for an offline course package.
// Mirrors CoursePackageManifest from the backend (Story 4.1 PKG-CONTRACT-2).
//
// Story 4.1 AC-1: defines course/version identifiers, checksums, size,
// effective time, files, minimum client version, and licenses.

import 'package:equatable/equatable.dart';

import 'package_content_type.dart';
import 'package_file_entry.dart';
import 'package_license.dart';

/// Accuracy class for package data quality.
enum AccuracyClass {
  aRtkSurveyed('A_RTK_SURVEYED'),
  bLicensedProvider('B_LICENSED_PROVIDER'),
  cVerifiedSatellite('C_VERIFIED_SATELLITE'),
  dUnverifiedCommunity('D_UNVERIFIED_COMMUNITY');

  final String value;
  const AccuracyClass(this.value);

  static AccuracyClass? fromString(String? value) {
    if (value == null) return null;
    return AccuracyClass.values.cast<AccuracyClass?>().firstWhere(
      (e) => e!.value == value,
      orElse: () => null,
    );
  }
}

/// Tiles format used in this package.
enum TilesFormat {
  pmtiles('PMTILES'),
  mbtiles('MBTILES'),
  vectorTiles('VECTOR_TILES');

  final String value;
  const TilesFormat(this.value);

  static TilesFormat fromString(String value) {
    return TilesFormat.values.firstWhere(
      (e) => e.value == value.toUpperCase(),
      orElse: () => TilesFormat.pmtiles,
    );
  }
}

/// Course package manifest — the offline package contract.
class CoursePackageManifest extends Equatable {
  final String packageId;
  final int courseId;
  final String version;
  final DateTime effectiveDate;
  final DateTime? expiresAt;
  final String checksum;
  final int sizeBytes;
  final TilesFormat tilesFormat;
  final String tilesUrl;
  final String geoJsonUrl;
  final String? dataVersion;
  final List<PackageFileEntry> files;
  final String minimumClientVersion;
  final List<PackageLicense> licenses;
  final String? scorecardUrl;
  final String? rulesUrl;
  final String? conditionsUrl;
  final String? metadataUrl;
  final DateTime generatedAt;
  final String generatedBy;
  final DateTime? pinSnapshotDate;
  final DateTime? weatherSnapshotDate;
  final AccuracyClass? accuracyClass;
  final double? confidence;

  const CoursePackageManifest({
    required this.packageId,
    required this.courseId,
    required this.version,
    required this.effectiveDate,
    this.expiresAt,
    required this.checksum,
    required this.sizeBytes,
    required this.tilesFormat,
    required this.tilesUrl,
    required this.geoJsonUrl,
    this.dataVersion,
    required this.files,
    required this.minimumClientVersion,
    required this.licenses,
    this.scorecardUrl,
    this.rulesUrl,
    this.conditionsUrl,
    this.metadataUrl,
    required this.generatedAt,
    required this.generatedBy,
    this.pinSnapshotDate,
    this.weatherSnapshotDate,
    this.accuracyClass,
    this.confidence,
  });

  /// Parse from API response JSON.
  factory CoursePackageManifest.fromJson(Map<String, dynamic> json) {
    return CoursePackageManifest(
      packageId: json['packageId'] as String,
      courseId: int.parse(json['courseId'].toString().split('/').last),
      version: json['version'] as String,
      effectiveDate: DateTime.parse(json['effectiveDate'] as String),
      expiresAt: json['expiresAt'] != null
          ? DateTime.parse(json['expiresAt'] as String)
          : null,
      checksum: json['checksum'] as String,
      sizeBytes: json['sizeBytes'] as int,
      tilesFormat: TilesFormat.fromString(json['tilesFormat'] as String),
      tilesUrl: json['tilesUrl'] as String,
      geoJsonUrl: json['geoJsonUrl'] as String,
      dataVersion: json['dataVersion'] as String?,
      files:
          (json['files'] as List<dynamic>?)
              ?.map((e) => PackageFileEntry.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      minimumClientVersion: json['minimumClientVersion'] as String,
      licenses:
          (json['licenses'] as List<dynamic>?)
              ?.map((e) => PackageLicense.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      scorecardUrl: json['scorecardUrl'] as String?,
      rulesUrl: json['rulesUrl'] as String?,
      conditionsUrl: json['conditionsUrl'] as String?,
      metadataUrl: json['metadataUrl'] as String?,
      generatedAt: DateTime.parse(json['generatedAt'] as String),
      generatedBy: json['generatedBy'] as String,
      pinSnapshotDate: json['pinSnapshotDate'] != null
          ? DateTime.parse(json['pinSnapshotDate'] as String)
          : null,
      weatherSnapshotDate: json['weatherSnapshotDate'] != null
          ? DateTime.parse(json['weatherSnapshotDate'] as String)
          : null,
      accuracyClass: AccuracyClass.fromString(json['accuracyClass'] as String?),
      confidence: (json['confidence'] as num?)?.toDouble(),
    );
  }

  Map<String, dynamic> toJson() => {
    'packageId': packageId,
    'courseId': courseId,
    'version': version,
    'effectiveDate': effectiveDate.toIso8601String(),
    if (expiresAt != null) 'expiresAt': expiresAt!.toIso8601String(),
    'checksum': checksum,
    'sizeBytes': sizeBytes,
    'tilesFormat': tilesFormat.value,
    'tilesUrl': tilesUrl,
    'geoJsonUrl': geoJsonUrl,
    if (dataVersion != null) 'dataVersion': dataVersion,
    'files': files.map((f) => f.toJson()).toList(),
    'minimumClientVersion': minimumClientVersion,
    'licenses': licenses.map((l) => l.toJson()).toList(),
    if (scorecardUrl != null) 'scorecardUrl': scorecardUrl,
    if (rulesUrl != null) 'rulesUrl': rulesUrl,
    if (conditionsUrl != null) 'conditionsUrl': conditionsUrl,
    if (metadataUrl != null) 'metadataUrl': metadataUrl,
    'generatedAt': generatedAt.toIso8601String(),
    'generatedBy': generatedBy,
    if (pinSnapshotDate != null)
      'pinSnapshotDate': pinSnapshotDate!.toIso8601String(),
    if (weatherSnapshotDate != null)
      'weatherSnapshotDate': weatherSnapshotDate!.toIso8601String(),
    if (accuracyClass != null) 'accuracyClass': accuracyClass!.value,
    if (confidence != null) 'confidence': confidence,
  };

  /// True if this package is currently within its effective date window.
  bool get isEffective {
    final now = DateTime.now();
    return effectiveDate.isBefore(now) &&
        (expiresAt == null || expiresAt!.isAfter(now));
  }

  @override
  List<Object?> get props => [
    packageId,
    courseId,
    version,
    effectiveDate,
    expiresAt,
    checksum,
    sizeBytes,
    tilesFormat,
    tilesUrl,
    geoJsonUrl,
    dataVersion,
    files,
    minimumClientVersion,
    licenses,
    scorecardUrl,
    rulesUrl,
    conditionsUrl,
    metadataUrl,
    generatedAt,
    generatedBy,
    pinSnapshotDate,
    weatherSnapshotDate,
    accuracyClass,
    confidence,
  ];
}

File: apps/api/src/main/java/vnpt/vsp/module/package/entity/CoursePackageManifest.java
package vnpt.vsp.module.pkg.entity;

import jakarta.persistence.*;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

/**
 * JPA entity for a course package manifest.
 *
 * Represents the offline course package contract (Story 4.1 AC-1).
 * Distinct from the booking/commercial CoursePackage in course.yaml.
 */
@Entity
@Table(name = "course_package_manifest",
    indexes = {
        @Index(name = "idx_manifest_course_version", columnList = "courseId, version", unique = true),
        @Index(name = "idx_manifest_course_effective", columnList = "courseId, effectiveFrom")
    })
public class CoursePackageManifest {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @NotNull
    @Column(nullable = false)
    private Long courseId;

    @NotNull
    private Long dataVersionId;

    @NotBlank
    @Column(nullable = false)
    private String version; // semver

    @NotNull
    @Min(0)
    @Column(nullable = false)
    private Long packageSizeBytes;

    @NotBlank
    @Column(nullable = false)
    private String checksum; // SHA-256 hex

    @NotNull
    @Column(nullable = false)
    private Instant effectiveFrom;

    private Instant expiresAt;

    @NotBlank
    @Column(nullable = false)
    private String minimumClientVersion; // semver

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private TilesFormat tilesFormat;

    @NotBlank
    @Column(nullable = false)
    private String tilesUrl;

    @NotBlank
    @Column(nullable = false)
    private String geoJsonUrl;

    private String scorecardUrl;
    private String rulesUrl;
    private String conditionsUrl;
    private String metadataUrl;

    @NotNull
    @Column(nullable = false)
    private Instant generatedAt;

    @NotBlank
    @Column(nullable = false)
    private String generatedBy;

    @Enumerated(EnumType.STRING)
    private AccuracyClass accuracyClass;

    private Double confidence;

    private Instant pinSnapshotDate;
    private Instant conditionsSnapshotDate;
    private Instant weatherSnapshotDate;

    @OneToMany(mappedBy = "manifest", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<PackageFileEntry> files = new ArrayList<>();

    @OneToMany(mappedBy = "manifest", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<PackageLicense> licenses = new ArrayList<>();

    public enum TilesFormat {
        PMTILES, MBTILES, VECTOR_TILES
    }

    public enum AccuracyClass {
        A_RTK_SURVEYED, B_LICENSED_PROVIDER, C_VERIFIED_SATELLITE, D_UNVERIFIED_COMMUNITY
    }

    // JPA-required no-arg constructor
    protected CoursePackageManifest() {}

    public CoursePackageManifest(Long courseId, Long dataVersionId, String version,
            Long packageSizeBytes, String checksum, Instant effectiveFrom,
            String minimumClientVersion, TilesFormat tilesFormat,
            String tilesUrl, String geoJsonUrl, Instant generatedAt, String generatedBy) {
        this.courseId = courseId;
        this.dataVersionId = dataVersionId;
        this.version = version;
        this.packageSizeBytes = packageSizeBytes;
        this.checksum = checksum;
        this.effectiveFrom = effectiveFrom;
        this.minimumClientVersion = minimumClientVersion;
        this.tilesFormat = tilesFormat;
        this.tilesUrl = tilesUrl;
        this.geoJsonUrl = geoJsonUrl;
        this.generatedAt = generatedAt;
        this.generatedBy = generatedBy;
    }

    // Getters
    public UUID getId() { return id; }
    public Long getCourseId() { return courseId; }
    public Long getDataVersionId() { return dataVersionId; }
    public String getVersion() { return version; }
    public Long getPackageSizeBytes() { return packageSizeBytes; }
    public String getChecksum() { return checksum; }
    public Instant getEffectiveFrom() { return effectiveFrom; }
    public Instant getExpiresAt() { return expiresAt; }
    public String getMinimumClientVersion() { return minimumClientVersion; }
    public TilesFormat getTilesFormat() { return tilesFormat; }
    public String getTilesUrl() { return tilesUrl; }
    public String getGeoJsonUrl() { return geoJsonUrl; }
    public String getScorecardUrl() { return scorecardUrl; }
    public String getRulesUrl() { return rulesUrl; }
    public String getConditionsUrl() { return conditionsUrl; }
    public String getMetadataUrl() { return metadataUrl; }
    public Instant getGeneratedAt() { return generatedAt; }
    public String getGeneratedBy() { return generatedBy; }
    public AccuracyClass getAccuracyClass() { return accuracyClass; }
    public Double getConfidence() { return confidence; }
    public Instant getPinSnapshotDate() { return pinSnapshotDate; }
    public Instant getConditionsSnapshotDate() { return conditionsSnapshotDate; }
    public Instant getWeatherSnapshotDate() { return weatherSnapshotDate; }
    public List<PackageFileEntry> getFiles() { return files; }
    public List<PackageLicense> getLicenses() { return licenses; }

    // Setters
    public void setExpiresAt(Instant expiresAt) { this.expiresAt = expiresAt; }
    public void setScorecardUrl(String scorecardUrl) { this.scorecardUrl = scorecardUrl; }
    public void setRulesUrl(String rulesUrl) { this.rulesUrl = rulesUrl; }
    public void setConditionsUrl(String conditionsUrl) { this.conditionsUrl = conditionsUrl; }
    public void setMetadataUrl(String metadataUrl) { this.metadataUrl = metadataUrl; }
    public void setAccuracyClass(AccuracyClass accuracyClass) { this.accuracyClass = accuracyClass; }
    public void setConfidence(Double confidence) { this.confidence = confidence; }
    public void setPinSnapshotDate(Instant pinSnapshotDate) { this.pinSnapshotDate = pinSnapshotDate; }
    public void setConditionsSnapshotDate(Instant conditionsSnapshotDate) { this.conditionsSnapshotDate = conditionsSnapshotDate; }
    public void setWeatherSnapshotDate(Instant weatherSnapshotDate) { this.weatherSnapshotDate = weatherSnapshotDate; }

    public void addFile(PackageFileEntry file) {
        files.add(file);
        file.setManifest(this);
    }

    public void addLicense(PackageLicense license) {
        licenses.add(license);
        license.setManifest(this);
    }
}
